### ISSUE TITLE
Cart session remains when changing user -  previous ticket #8504

### ISSUE THREAD

**Issue by PyramidMDT** [woocommerce/woocommerce#26359]
Hello
Wordpress + Woocommerace
Hosting - Siteground Business
Payment - Stripe

Customers are seeing sender and recipient information populating random customers on other machines. Also random items are also being populated in random carts. 
Nginx cach is turned off 
Rules to block cart, account and checkout page from caching is turned on

Is anyone else experiencing this phenomenon as well?

**Describe the bug**
A clear and concise description of what the bug is. Please be as descriptive as possible; issues lacking detail, or for any other reason than to report a bug, may be closed without action.

This checkout for should not be keeping information and then pushing it up to the browser from a different session. 

All software is lates version WP = 5.4.1. Woo 4.0.1
Plug ins
Checkout Field editor 3.0.8
Contact Form 7 5.1.7
Facebook for Woo 1.11.1
Flexible product fields - 1.2.16
Jet Pack - 8.4.2 -not activated
Limit Login attempts Reloaded 2.12.3
Mailchimp for WooCommerce 1.0.0
MAS API Collection 1.0.0
Nextend Social Login 3.0.22
SG Optimizer 5.5.3
WooCommerce 4.0..1
WooCommerce Continue Shopping 1.4.2
WooCommerce Services 1.23.0
WooCommerce Stripe Gateway 4.3.3
WooSidebars 1.4.5

`
### WordPress Environment ###

WordPress address (URL): https://hoelcks.com
Site address (URL): https://hoelcks.com
WC Version: 4.0.1
REST API Version: ✔ 1.0.7
WC Blocks Version: ✔ 2.5.14
Action Scheduler Version: ✔ 3.1.4
WC Admin Version: ✔ 1.0.3
Log Directory Writable: ✔
WP Version: 5.4.1
WP Multisite: –
WP Memory Limit: 768 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache
PHP Version: 7.3.16
PHP Post Max Size: 128 MB
PHP Time Limit: 120
PHP Max Input Vars: 3000
cURL Version: 7.66.0
OpenSSL/1.1.1d-fips

SUHOSIN Installed: –
MySQL Version: 5.7.25-28-log
Max Upload Size: 128 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 4.0.1
WC Database Prefix: wp_
Total Database Size: 6.31MB
Database Data Size: 5.44MB
Database Index Size: 0.87MB
wp_woocommerce_sessions: Data: 1.54MB + Index: 0.04MB + Engine MyISAM
wp_woocommerce_api_keys: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_attribute_taxonomies: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_downloadable_product_permissions: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_order_items: Data: 0.01MB + Index: 0.01MB + Engine MyISAM
wp_woocommerce_order_itemmeta: Data: 0.07MB + Index: 0.06MB + Engine MyISAM
wp_woocommerce_tax_rates: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_woocommerce_tax_rate_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zones: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zone_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zone_methods: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_payment_tokens: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_payment_tokenmeta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_actionscheduler_actions: Data: 0.04MB + Index: 0.03MB + Engine MyISAM
wp_actionscheduler_claims: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_actionscheduler_groups: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_actionscheduler_logs: Data: 0.03MB + Index: 0.02MB + Engine MyISAM
wp_commentmeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_comments: Data: 0.02MB + Index: 0.02MB + Engine MyISAM
wp_links: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_mailchimp_carts: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_mailchimp_jobs: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_options: Data: 2.05MB + Index: 0.10MB + Engine MyISAM
wp_postmeta: Data: 0.65MB + Index: 0.33MB + Engine MyISAM
wp_posts: Data: 0.97MB + Index: 0.08MB + Engine MyISAM
wp_social_users: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_termmeta: Data: 0.01MB + Index: 0.01MB + Engine MyISAM
wp_terms: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_term_relationships: Data: 0.01MB + Index: 0.02MB + Engine MyISAM
wp_term_taxonomy: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_usermeta: Data: 0.02MB + Index: 0.02MB + Engine MyISAM
wp_users: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_admin_notes: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_admin_note_actions: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_category_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_customer_lookup: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_download_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_order_coupon_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_order_product_lookup: Data: 0.01MB + Index: 0.01MB + Engine MyISAM
wp_wc_order_stats: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_order_tax_lookup: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_product_meta_lookup: Data: 0.01MB + Index: 0.02MB + Engine MyISAM
wp_wc_tax_rate_classes: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_webhooks: Data: 0.00MB + Index: 0.00MB + Engine MyISAM

### Post Type Counts ###

attachment: 108
blocks: 9
customize_changeset: 20
featured_item: 8
fpf_fields: 1
nav_menu_item: 77
page: 85
post: 7
product: 55
product_variation: 135
revision: 206
shop_coupon: 1
shop_order: 64
wcpa_pt_forms: 1
wpcf7_contact_form: 5

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (15) ###

MAS API Collection: by MAS Direct Network – 1.0.0
Contact Form 7: by Takayuki Miyoshi – 5.1.7
WooCommerce Continue Shopping: by HappyKite – 1.4.2 – Not tested with the active version of WooCommerce
Facebook for WooCommerce: by Facebook – 1.11.1 – Not tested with the active version of WooCommerce
Flexible Product Fields: by WP Desk – 1.2.16 – Not tested with the active version of WooCommerce
Jetpack by WordPress.com: by Automattic – 8.4.2
Limit Login Attempts Reloaded: by WPChef – 2.12.3
Mailchimp for WooCommerce: by Mailchimp – 2.4.0 – Not tested with the active version of WooCommerce
Nextend Social Login: by Nextendweb – 3.0.22
SG Optimizer: by SiteGround – 5.5.3
Checkout Field Editor for WooCommerce: by ThemeHigh – 3.0.8 – Not tested with the active version of WooCommerce
WooCommerce Stripe Gateway: by WooCommerce – 4.3.3 – Not tested with the active version of WooCommerce
WooCommerce Services: by Automattic – 1.23.0 – Not tested with the active version of WooCommerce
WooCommerce: by Automattic – 4.0.1
WooSidebars: by WooCommerce – 1.4.5

### Inactive Plugins (1) ###

Akismet Anti-Spam: by Automattic – 4.1.5

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #64 - /shop/
Cart: #63 - /cart/
Checkout: #62 - /checkout/
My account: #61 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Flatsome
Version: 3.11.3
Author URL: http://www.uxthemes.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: flatsome/woocommerce/archive-product.php
flatsome/woocommerce/cart/cart-empty.php
flatsome/woocommerce/cart/cart-shipping.php
flatsome/woocommerce/cart/cart.php
flatsome/woocommerce/checkout/form-checkout.php
flatsome/woocommerce/checkout/form-coupon.php
flatsome/woocommerce/checkout/terms.php
flatsome/woocommerce/checkout/thankyou.php
flatsome/woocommerce/content-product.php
flatsome/woocommerce/content-product_cat.php
flatsome/woocommerce/content-single-product.php
flatsome/woocommerce/content-widget-product.php
flatsome/woocommerce/global/breadcrumb.php
flatsome/woocommerce/global/quantity-input.php
flatsome/woocommerce/global/wrapper-end.php
flatsome/woocommerce/global/wrapper-start.php
flatsome/woocommerce/loop/loop-end.php
flatsome/woocommerce/loop/loop-start.php
flatsome/woocommerce/loop/pagination.php
flatsome/woocommerce/loop/result-count.php
flatsome/woocommerce/loop/sale-flash.php
flatsome/woocommerce/myaccount/form-login.php
flatsome/woocommerce/myaccount/navigation.php
flatsome/woocommerce/notices/error.php
flatsome/woocommerce/notices/notice.php
flatsome/woocommerce/notices/success.php
flatsome/woocommerce/product-searchform.php
flatsome/woocommerce/single-product/photoswipe.php
flatsome/woocommerce/single-product/price.php
flatsome/woocommerce/single-product/product-image.php
flatsome/woocommerce/single-product/product-thumbnails.php
flatsome/woocommerce/single-product/rating.php
flatsome/woocommerce/single-product/related.php
flatsome/woocommerce/single-product/review.php
flatsome/woocommerce/single-product/sale-flash.php
flatsome/woocommerce/single-product/share.php
flatsome/woocommerce/single-product/short-description.php
flatsome/woocommerce/single-product/tabs/tabs.php
flatsome/woocommerce/single-product/title.php
flatsome/woocommerce/single-product/up-sells.php
flatsome/woocommerce/single-product-reviews.php
flatsome/woocommerce/single-product.php

### Action Scheduler ###

Complete: 155
Oldest: 2020-04-22 13:41:02 +0000
Newest: 2020-05-01 18:09:47 +0000

Thank you 
Thor Torgersen

**Comment by juliaamosova** [comment]
Hi @PyramidMDT, thanks for reporting the issue. Linking the previous thread here for reference: https://github.com/woocommerce/woocommerce/issues/8504.

**Comment by doubleo2** [comment]
I have the same issue. Google turns up a surprising number of people reporting this issue, but no solutions 😞

**Comment by doubleo2** [comment]
@PyramidMDT I've personally never seen it (unfortunately), but I've had at least two reports from customers in the past week. But that obviously doesn't include anyone who saw it and left because they couldn't trust us to keep their information secure.

Given that people have been reporting this for a few years now I suspect it's not related to any plugins, but just in case I compared the list of plugins in your bug report with the active plugins on my site and there are only four we are both using.

```
Facebook for WooCommerce: by Facebook – 1.10.2
Mailchimp for WooCommerce: by Mailchimp – 2.3.6
WooCommerce Stripe Gateway: by WooCommerce – 4.3.3
WooCommerce: by Automattic – 4.0.1
```

**Comment by doubleo2** [comment]
@PyramidMDT Have you seen it yourself? Any ideas on how to reproduce it? The two users who reported both arrived on my site via referral, but that may just be a coincidence.

**Comment by PyramidMDT** [comment]
I have seen it myself, both with products being added and pre-population of checkout page with other customers info shipping and billing address. It appears to be a caching issue and siteground is saying they are not caching. Ihave turned off all caching I can find, but it just happened again about 10 min ago to another client. Only hearing about it if customer reports it so I have no idea how often it may be happening. Going to implement google event tracking and see how many bailing at checkout.

**Comment by PyramidMDT** [comment]
@doubleo2 I am not running facebook or mailchimp. Are you hosting on siteground by chance?

**Comment by doubleo2** [comment]
@PyramidMDT I am not using Mailchimp either but I still have the plugin installed. No, my site is hosted on Kinsta. I have also verified that the cart and checkout pages are not being cached.

**Comment by doubleo2** [comment]
I should probably also note that I don't have user accounts enabled. All customers are anonymous users - i.e. logged out. I have also verified that all pages on my site bypass the cache once I've added an item to my cart, which is how Kinsta describes their caching as working.

> Certain WooCommerce and Easy Digital Downloads pages like cart, my-account, and checkout, are automatically excluded from caching. There is a server-level rule in place so that users automatically bypass the cache when the woocommerce_items_in_cart cookie or edd_items_in_cart cookie is detected to ensure a smooth and in-sync checkout process.
> 
> We also listen for the associated logged-in cookies and set the cache to bypass when we detect that someone has logged into WordPress. The prevents the back-end dashboard from accidentally being cached.
> 
> By default, we don’t exclude the wp_woocommerce_session_ cookie from caching. Most WooCommerce sites in our experience don’t have any issues. This also improves performance by increasing your cache HIT ratio, while utilizing fewer PHP workers.

**Comment by doubleo2** [comment]
As a stop gap measure I have modified `form-billing.php` and `form-shipping.php` in `wp-content/plugins/woocommerce/templates/checkout/` as follows

```php
# form-billing.php#L39

foreach ( $fields as $key => $field ) {
    # REPLACE woocommerce_form_field( $key, $field, $checkout->get_value( $key ) );
    woocommerce_form_field( $key, $field, '' );
}
```

```php
# form-shipping#L39

foreach ( $fields as $key => $field ) {
    # REPLACE woocommerce_form_field( $key, $field, $checkout->get_value( $key ) );
    woocommerce_form_field( $key, $field, '' );
}

# form-shipping.php#L63

<div class="woocommerce-additional-fields__field-wrapper">
    <?php foreach ( $checkout->get_checkout_fields( 'order' ) as $key => $field ) : ?>
        <!-- <?php # REPLACE woocommerce_form_field( $key, $field, $checkout->get_value( $key ) ); ?>-->
        <?php woocommerce_form_field( $key, $field, '' ); ?>
    <?php endforeach; ?>
</div>
```

I believe that will at least prevent WooCommerce from leaking customer data until this issue is addressed.

**Comment by Robbertdk** [comment]
Yes, I have this same issue for a couple of weeks and just can't find out what causes it. I can't replicate the problem myself, we see this problem coming back sporadicly, once a couple of days.

Not running on Siteground either, page is not cached.

**Comment by Robbertdk** [comment]
Interesting: just like @doubleo2 I have a multsite with MultilangualPress setup.

**Comment by PyramidMDT** [comment]
Happened again this morning to client. I have not tried the code provided by doublo2 yet. 
Is it possible it is caching even when you have it turned off?

**Comment by peterfabian** [comment]
Hi all, thanks for reporting the issue.

As I was looking at the respective code, I was checking previous issues and found this one: https://github.com/woocommerce/woocommerce/issues/18561

Are all of you using Stripe payment gateway? We might need to check if we aren't clashing somewhere with our session handling. 

I'm trying to find out where there might be a clash happening, but all I see is only potential spots for issues introduced by caching (at least with WC core code without extensions).

When a customer adds something to the cart, WC core generates a unique customer_id for this customer, based on random 32 bytes from `/dev/urandom`, which then goes through `md5` function. 
This customer_id is then used as an identifier in the database to retrieve the session information. However, the session information is only retrieved if the cookie expiration time + customer_id hash matches the hash stored in the cookie itself. 
If there is a new visitor/customer who just arrived at the site, they would have to receive the same random `$customer_id`, or, they would have to have received it at some point in the past (along with a valid cookie expiry that hasn't passed yet).

Caching on the WordPress level uses the cache key that includes `$customer_id`, so again, as long as random number generator is not broken, the data from another customer should not get loaded. 

One idea would be to check the `wp_woocommerce_sessions` db table to see if there isn't anything weird happening there.

I'm not sure if possible, but maybe you could add logging to the function `generate_customer_id` with a timestamp to see if it is indeed the broken random generator. 
 
Another idea would be to add logging to `wp_cache_get` and `wp_cache_set` for group `wc_session_id` and try to correlate it with the reports from customers. 

Maybe someone else has a better idea, but as long as we're unable to replicate this problem, some detective work and reasoning about code seems like one of the few possibilities we have.

**Comment by Robbertdk** [comment]
Thanks @peterfabian. Not using the Stripe gateway, but I did found some keys that looked different than other keys. 

![looked different than other keys](https://user-images.githubusercontent.com/4647242/81947322-b475b200-9600-11ea-8fc3-10c33be8a26f.png)

When looking in the codebase I found that the plugin WooCommerce EU Vat Compliance also sets a cookie via `set_customer_session_cookie` just as the Stripe gateway does. Don't know if that causes a problem, though.

**Comment by peterfabian** [comment]
Thanks for your update. I think those numeric keys (10228, 10230) relate to your customers who have an account while the alphanumeric ones are guest sessions. I'll try to check the code of the EU Vat plugin to see if there is anything suspicious there.

**Comment by nine2uk** [comment]
Also sad to report we're facing this issue on two separate sites at the moment. 

Both using Stripe and both hosted with Kinsta.

**Comment by peterfabian** [comment]
For those of you not using MailChimp, can you try to disable to plugin to see if it makes any difference please?

**Comment by doubleo2** [comment]
FWIW I believe my customers still experienced this issue after I uninstalled MailChimp. I have not received any new reports since applying my workaround - but it's a hack not a fix.

**Comment by peterfabian** [comment]
Anyone able to add more logging to their server to around the places I mentioned? Otherwise, we'd probably close the issue as not reproducible, since it's not clear how this could be happening. Thanks.

**Comment by Robbertdk** [comment]
No, but I've implemented te hack from @doubleo2, forgot about it and with the last WooCommerce update I've overwritten the changed files and thereby the hack. But, after that I've haven't heard of new cases so it seems like it was resolved in the update or stopped somehow... weird bug.

**Comment by peterfabian** [comment]
Thanks for reporting back. I think I'll close this until we get more specific information. Otherwise, there is no real way for us to fix this unless we have more information/way to reproduce. Feel free to comment if you get more info, we can reopen if needed.

**Comment by CaptnAbap** [comment]
We've just update to the newest woocommerce version (and update all other plugins). That did not help. And it happens currently every single time somebody buys something.
@peterfabian we have a test environment, do you want to see it there?

**Comment by peterfabian** [comment]
Sure, if possible, please create an admin account for woologin@woocommerce.com and let us know your URL so that we can log in and check what is happening.

**Comment by CaptnAbap** [comment]
Hi Peter

We've created you an account on the staging environment.
http://staging3.sato-code.com/
with
woologin@woocommerce.com
and pw
5^0G9VGEIM76i@1yEFv$a!pI

Following Testcase:
1. Buy a product in a Language. There is currently the stripe test key.
    We used for buying the product Safari on an iPhone
2. Come back to the page (same language) by using a different browser and
the checkout fields will be filled.
    I've tested a couple of browser and it does not always appear. Last
time it showed on Edge and Chrome.

Let me know if you need anything.

Best
Stefan
halterstefan@gmail.com

**Comment by CaptnAbap** [comment]
We've created you an account on the staging environment.
http://staging3.sato-code.com/
with
woologin@woocommerce.com  
drop me pls an email for the password: stefan.halter@sato-code.com

Following Testcase:
1. Buy a product in a Language. There is currently the stripe test key.
    We used for buying the product Safari on an iPhone
2. Come back to the page (same language) by using a different browser and the checkout fields will be filled. 
    I've tested a couple of browser and it does not always appear. Last time it showed on Edge and Chrome.

**Comment by CaptnAbap** [comment]
If this is really a cookie problem. Did somebody try to set the woocommerce session duration to 15min? This would at least decrease the chance that it happens. 

add_filter('wc_session_expiring', 'filter_ExtendSessionExpiring' );
add_filter('wc_session_expiration' , 'filter_ExtendSessionExpired' );

function filter_ExtendSessionExpiring($seconds) {
    return 60 * 15;
}
function filter_ExtendSessionExpired($seconds) {
   return 60 * 15;
}

**Comment by peterfabian** [comment]
Hi @CaptnAbap , 

Just tried it now, bought Bellinzona Trail   × 1, checked out in Chrome, switched to Firefox and I can't see any data pre-filled. Same with Safari, no info. 
Then tried checking out in Firefox, switching to other browsers, I still see no info getting leaked.

The data is only pre-filled if I try again in the same browser (ie checkout in Chrome, then navigate somewhere else, go back to your site and go to checkout, but I assume that's not what we're trying to solve now).


![Screenshot 2020-08-28 at 15 23 44](https://user-images.githubusercontent.com/2207451/91565451-82cfef00-e942-11ea-9d82-7c2e52b28525.png)

I don't think it could be a problem with cookies timeout. The backend (PHP script) that serves cookies would have to identify the user as being the same a user coming from another browser. Unless I missed something, that can only happen if 
- a) those users share cookies between each other--highly unlikely. 
- b) random number generator assigned the same random string to 2 consecutive visits--again, very unlikely. 

That being said, if you are able to intermittently trigger this problem, I think you could try to add more logging to your staging site (e.g. log time, IP address, user agent and cookie that's getting sent out by `wc_setcookie`) and take a note of time when you see it happening to be able to identify the spot in the log file when it happened?

**Comment by CaptnAbap** [comment]
Yes, when we saw that you bought something we tried to see the checkout information as well. But nothing was displayed. The only thing that was different now was that we had the test key in stripe. 

However, I saw myself information that was entered on another computer. So, it happens somehow. But it's hard to get the right constellation to trigger it.

**Comment by CaptnAbap** [comment]
So, have it. Go to the website: staging3.sato-code.com/?lang=de
select the Product Albert's Door and you will see the customer details. 

I guess to trigger it, it needs to be in another language. Not the main language of the website.

**Comment by CaptnAbap** [comment]
![grafik](https://user-images.githubusercontent.com/31867037/91582767-8c605380-e950-11ea-9bc2-05232f7838f0.png)

The product was bought in German with a safari browser on an iPhone. This screenshot was taken on an Edge on a Laptop.

**Comment by peterfabian** [comment]
Sorry, I can't get it triggered even with the German language. Though I'm a bit suspicious if some of the plugins you have activated isn't causing it. Have you managed to capture anything useful in the logs?

**Comment by CaptnAbap** [comment]
There are two issues.

[28-Aug-2020 09:23:46 UTC] PHP Fatal error:  Uncaught Error: Call to a member function get_type() on bool in /home/customer/www/staging3.XXXXXXXXXXXX.com/public_html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php:1012
Stack trace:
#0 /home/customer/www/staging3.XXXXXXXXXXXX.com/public_html/wp-includes/class-wp-hook.php(287): WC_Stripe_Payment_Request->ajax_add_to_cart('')
#1 /home/customer/www/staging3.XXXXXXXXXXXX.com/public_html/wp-includes/class-wp-hook.php(311): WP_Hook->apply_filters('', Array)
#2 /home/customer/www/staging3.XXXXXXXXXXXX.com/public_html/wp-includes/plugin.php(478): WP_Hook->do_action(Array)
#3 /home/customer/www/staging3.XXXXXXXXXXXX.com/public_html/wp-content/plugins/woocommerce/includes/class-wc-ajax.php(89): do_action('wc_ajax_wc_stri...')
#4 /home/customer/www/staging3.XXXXXXXXXXXX.com/public_html/wp-includes/class-wp-hook.php(287): WC_AJAX::do_wc_ajax('')
#5 /home/customer/www/staging3.XXXXXXXXXXXX.com/public_html/wp-includes/class-wp-hook.php(311): WP_Hook->apply_filters(false, Array)
#6 in /home/customer/www/staging3.XXXXXXXXXXXX.com/public_html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php on line 1012
[28-Aug-2020 09:29:38 UTC] The The wc_add_to_cart_message filter function is deprecated since version 3.0. Replace with wc_add_to_cart_message_html.

**Comment by CaptnAbap** [comment]
So, I've spent pretty much the whole day trying to track down this issue. As described above, I found a test case. With that I started changing configs and one seemed to work:
We're using siteground and the SG optimizer. When the SG optimizer is deactivated the problem did not occur. Since this is most likely a caching issue, it makes sense that this could be the root cause. So, I'm going ahead and turn off the super cacher of in the SG optimizer. 
I've also posted it to the siteground support. They could not reproduce it as well. Probably because it's a very specific constellation. 

It would be great if others could also share findings. This issue is really hard to get a grip on.

**Comment by CaptnAbap** [comment]
So, we got it finally figured out. It was a server-side caching issue. 

We use siteground and with it comes the SG optimizer. The SG optimizer caches complete websites. This leads to the problem that a second user get's the same website as the previous user.
To solve it make the following steps:
In wordpress, in the plugin SG optimizer, there is a way to exclude some URLs  from caching. In our case we excluced /cart/* and /checkout/* from being cached. Here is the manual for it: https://www.siteground.com/tutorials/wordpress/sg-optimizer/supercacher/

Secondly, I'd recommend to exclude the same pages to be excluded from cloudflare caching. It did not seem to be the problem. But let's do it properly: Go to cloudflare and make a page rule for the same URLs (/cart/* and /checkout/*). I recommend a page rule on cache level = Bypass. Here is the manual for that:

https://support.cloudflare.com/hc/en-us/articles/218411427-Understanding-and-Configuring-Cloudflare-Page-Rules-Page-Rules-Tutorial-

I hope that helps, at least if you're hosting with siteground.
